cpp/src/arrow/engine/substrait/registry.h

```diff
@@ -0,0 +1,86 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// NOTE: API is EXPERIMENTAL and will change without going through a
+// deprecation cycle
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "arrow/result.h"
+#include "arrow/status.h"
+#include "arrow/util/visibility.h"
+
+#include "arrow/compute/exec/exec_plan.h"
+#include "arrow/engine/substrait/extension_set.h"
+#include "arrow/engine/substrait/extension_types.h"
+#include "arrow/engine/substrait/options.h"
+#include "arrow/engine/substrait/relation_internal.h"
+#include "arrow/engine/substrait/serde.h"
+#include "arrow/engine/substrait/visibility.h"
+#include "arrow/type_fwd.h"
+
+#include "substrait/algebra.pb.h"  // IWYU pragma: export
+
+namespace arrow {
+
+namespace engine {
+
+/// \brief Acero-Substrait integration contains converters which enables
+/// converting Acero ExecPlan related entities to the corresponding Substrait
+/// entities.
+///
+/// Note that the current registry definition only holds converters to convert
+/// an Acero plan to Substrait plan.
+class ARROW_ENGINE_EXPORT SubstraitConversionRegistry {
+ public:
+  virtual ~SubstraitConversionRegistry() = default;
+
+  /// \brief Alias for Acero-to-Substrait converter
+  using SubstraitConverter = std::function<Result<std::unique_ptr<substrait::Rel>>(
+      const std::shared_ptr<Schema>&, const compute::Declaration&, ExtensionSet*,
+      const ConversionOptions&)>;
```

Comment:
Since this class references the protobuf message classes, it should be confined to `_internal.h` and `.cc` files to ensure it is not installed along with the other arrow headers; the protobuf message classes are currently private/internal

Reply:
this is removed now.